File: backend/src/routes/webhook.js
const express = require('express');
const crypto  = require('crypto');
const db      = require('../db');
const events  = require('../events');
const router  = express.Router();

/**
 * OME Admission Webhook
 * OME calls this for every ingest connection to validate the stream key.
 *
 * Request body from OME:
 * {
 *   "client": { "address": "...", "port": ... },
 *   "request": {
 *     "direction": "incoming",   // "incoming" = ingest, "outgoing" = viewer
 *     "protocol": "rtmp" | "srt" | "webrtc",
 *     "url": "rtmp://host/live/stream-key-here",
 *     "time": "..."
 *   }
 * }
 *
 * We respond with:
 *   { "allowed": true }   → accept
 *   { "allowed": false }  → deny
 */
router.post('/', (req, res) => {
    // req.body is the raw Buffer (express.raw middleware is applied before this route)
    const rawBody   = req.body;
    const signature = req.headers['x-ome-signature'];

    if (!signature) {
        console.warn('[webhook] missing signature — rejected');
        return res.status(401).json({ allowed: false });
    }
    const expected = crypto
        .createHmac('sha1', process.env.OME_WEBHOOK_SECRET || '')
        .update(rawBody)
        .digest('base64url');
    if (signature !== expected) {
        console.warn('[webhook] signature mismatch — rejected');
        return res.status(401).json({ allowed: false });
    }

    let parsed;
    try { parsed = JSON.parse(rawBody); } catch {
        return res.status(400).json({ allowed: false });
    }

    const { request } = parsed;
    if (!request) return res.status(400).json({ allowed: false });

    const direction = request.direction; // 'incoming' = ingest, 'outgoing' = viewer
    const url       = request.url || '';

    if (direction === 'incoming') {
        // Extract stream key from URL
        // RTMP:  rtmp://host/live/{stream-key}
        // SRT:   streamid = default/live/{stream-key}  (OME resolves to URL)
        // WHIP:  https://host/live/{stream-key}?direction=whip
        const streamKey = extractStreamKey(url, request.protocol);

        if (!streamKey) {
            console.warn('[webhook] Could not extract stream key from:', url);
            return res.json({ allowed: false });
        }

        const key = db.prepare('SELECT * FROM stream_keys WHERE key_token = ?').get(streamKey);

        if (!key) {
            console.warn('[webhook] Unknown stream key:', streamKey);
            return res.json({ allowed: false });
        }

        // Capture slugs of rooms that are about to go live (so we can notify participants)
        const pendingSlugs = db.prepare(`
            SELECT slug FROM rooms WHERE stream_key_id = ? AND status = 'pending'
        `).all(key.id).map(r => r.slug);

        // Mark ALL rooms using this key as live
        const updated = db.prepare(`
            UPDATE rooms SET status = 'live', started_at = CURRENT_TIMESTAMP
            WHERE stream_key_id = ? AND status != 'ended'
        `).run(key.id);

        // Notify connected participants so their players start automatically
        for (const slug of pendingSlugs) events.emit('room:live', slug);

        console.log(`[webhook] Stream key ${streamKey} accepted → ${updated.changes} room(s) live`);
        return res.json({ allowed: true });
    }

    // 'outgoing' = viewer request (Phase 2: enforce waiting room here)
    // For now, allow all viewer connections — access controlled by the viewer join flow
    return res.json({ allowed: true });
});

function extractStreamKey(url, protocol) {
    try {
        // Normalise: SRT streamid comes through as a URL path, e.g. /default/live/{key}
        // RTMP and WHIP: last path segment after /live/
        const parsed = new URL(url.startsWith('http') || url.startsWith('rtmp') ? url : 'http://x' + url);
        const parts  = parsed.pathname.split('/').filter(Boolean);
        // Path is typically: [vhost,] app, streamKey — take the last segment
        return parts[parts.length - 1] || null;
    } catch {
        return null;
    }
}


module.exports = router;

File: backend/src/routes/webhook.rs
use axum::{body::Bytes, extract::State, http::HeaderMap, routing::post, Json, Router};
use base64::engine::general_purpose::URL_SAFE_NO_PAD;
use base64::Engine;
use hmac::{Hmac, Mac};
use rusqlite::params;
use serde_json::{json, Value};
use sha1::Sha1;
use std::sync::Arc;
use subtle::ConstantTimeEq;

use crate::error::AppError;
use crate::state::AppState;

type HmacSha1 = Hmac<Sha1>;

fn extract_stream_key(url: &str) -> Option<String> {
    let url_str = if url.starts_with("http") || url.starts_with("rtmp") {
        url.to_string()
    } else {
        format!("http://x{}", url)
    };
    let path = if let Some(pos) = url_str.find("://") {
        let after_scheme = &url_str[pos + 3..];
        after_scheme
            .find('/')
            .map(|p| &after_scheme[p..])
            .unwrap_or("")
    } else {
        &url_str
    };
    let parts: Vec<&str> = path.split('/').filter(|s| !s.is_empty()).collect();
    parts
        .last()
        .map(|s| s.split('?').next().unwrap_or(s).to_string())
}

async fn webhook_handler(
    State(state): State<Arc<AppState>>,
    headers: HeaderMap,
    body: Bytes,
) -> Result<Json<Value>, AppError> {
    // Verify HMAC signature
    let signature = headers
        .get("x-ome-signature")
        .and_then(|v| v.to_str().ok())
        .ok_or_else(|| AppError::Unauthorized("Missing signature".into()))?;

    let secret = &state.config.ome_webhook_secret;
    let mut mac = HmacSha1::new_from_slice(secret.as_bytes())
        .map_err(|e| AppError::Internal(format!("HMAC init error: {}", e)))?;
    mac.update(&body);
    let result = mac.finalize().into_bytes();
    let expected = URL_SAFE_NO_PAD.encode(result);

    if expected.as_bytes().ct_eq(signature.as_bytes()).unwrap_u8() == 0 {
        return Err(AppError::Unauthorized("Invalid signature".into()));
    }

    // Parse body as JSON
    let payload: Value =
        serde_json::from_slice(&body).map_err(|e| AppError::BadRequest(e.to_string()))?;

    let request = payload
        .get("request")
        .ok_or_else(|| AppError::BadRequest("Missing request object".into()))?;

    let direction = request
        .get("direction")
        .and_then(|v| v.as_str())
        .unwrap_or("");

    if direction != "incoming" {
        return Ok(Json(json!({ "allowed": true })));
    }

    // Extract stream key from URL
    let url = request.get("url").and_then(|v| v.as_str()).unwrap_or("");

    let stream_key = extract_stream_key(url)
        .ok_or_else(|| AppError::BadRequest("Cannot extract stream key".into()))?;

    // Look up stream key and update rooms
    let events = state.events.clone();
    let conn = state.db.get()?;
    let slugs = tokio::task::spawn_blocking(move || {
        // Find rooms with this stream key
        let mut stmt = conn.prepare(
            "SELECT r.id, r.slug FROM rooms r \
             JOIN stream_keys sk ON sk.id = r.stream_key_id \
             WHERE sk.key_token = ?1",
        )?;
        let rooms: Vec<(String, String)> = stmt
            .query_map(params![stream_key], |row| {
                Ok((row.get::<_, String>(0)?, row.get::<_, String>(1)?))
            })?
            .collect::<Result<Vec<_>, _>>()?;

        let mut slugs = Vec::new();
        for (room_id, slug) in &rooms {
            conn.execute(
                "UPDATE rooms SET status = 'live' WHERE id = ?1",
                params![room_id],
            )?;
            slugs.push(slug.clone());
        }
        Ok::<_, rusqlite::Error>(slugs)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))??;

    // Emit room:live events
    for slug in &slugs {
        let _ = events.room_live.send(slug.clone());
    }

    Ok(Json(json!({ "allowed": true })))
}

pub fn router() -> Router<Arc<AppState>> {
    Router::new().route("/", post(webhook_handler))
}
